### PR TITLE
update many docs to point at new locations and have better instructions

### PR DESCRIPTION
**Pull request by ahgittin**
these docs are also pushed to be live at the /v/0.9.0-SNAPSHOT/ endpoint

### CODE REVIEW

**Comment by ahgittin** [comment]
merging as they'll be much better than the old ones.  CTR request!  @rdowner you wrote the original so I'd esp like to know if this looks good to you, esp the way we now instruct for multi-repos, and how the org names here are changed

File: guide/dev/code/submodules.md [hunk 1]
```diff
@@ -5,31 +5,55 @@ layout: website-normal
...
 and push to your fork.
-You'll need to create a fork and the set up your local submodules should then have:
+You'll need to create a fork and then set up your local submodules should then have:
 
 * your fork as the `origin` remote repo
-* the Apache github repo as the `upstream` remote repo
-* optionally the repo at the Apache git server as the `apache` remote repo  
+* the Apache GitHub repo as the `upstream` remote repo, 
+  optionally with `origin` set as the `push` target (since you can't push to `github.com/apache`)
+  and fetch set up to get pull-request branches
```
Comment (rdowner) [inline guide/dev/code/submodules.md]: remove this line - only committers need PR branches and the "committers guide" in the website should describe how to set that up. For the majority of potential contributors, the PR branches are just noise.

File: guide/dev/code/submodules.md [hunk 2]
```diff
@@ -5,31 +5,55 @@ layout: website-normal
...
+* optionally the repo at the Apache git server as the `apache-git` remote repo
```
Comment (rdowner) [inline guide/dev/code/submodules.md]: same here

File: guide/dev/code/submodules.md [hunk 3]
```diff
@@ -5,31 +5,55 @@ layout: website-normal
...
+for x in brooklyn-* ; do
+  # repeat the above for each subproject
+  pushd $x
+  hub fork
+  git remote set-url --push upstream $(git remote -v | grep origin | grep push | awk '{print $2}')
+  git checkout master && git branch --set-upstream-to upstream/master
+  git config --local --add remote.upstream.fetch '+refs/pull/*/head:refs/remotes/upstream/pr/*'
+  git remote add apache-git https://git-wip-us.apache.org/repos/asf/$x
+  popd
+done
```
Comment (rdowner) [inline guide/dev/code/submodules.md]: Good grief that is complicated! Way too long to be in documentation. Suggest that this is a script in the root `brooklyn` repository. I also suggest that there are either separate scripts for non-committer developers and committers, or that the script takes a parameter to perform set up for committers (see earlier comments)

**Comment by rdowner** [comment]
I'm happy with the content.

However I think that this information should be in the "Developers" section of the website, and not part of the user guide (same comment applies to PR #1 on this project)

Reasons:
* The first place a would-be developer would look when wanting to get started is under the "Developers" heading on the main website. This information is not directly linked.
* Putting it in the user guide freezes it in time once a release is made. If the procedures change again in the future, the old version will permanently remain. Whereas by putting it in the website, always the latest information is permalinked and obsolete information removed.
* It currently only exists as a /v/0.9.0-SNAPSHOT/* URL which is an obscure URL (i.e. starting from brooklyn.apache.org it is many clicks to get there)

File: guide/dev/code/submodules.md [hunk 4]
```diff
@@ -5,31 +5,55 @@ layout: website-normal
...
-    git remote add apache https://git-wip-us.apache.org/repos/asf/$x
-    for x in brooklyn-* ; do
-      pushd $x
-      hub fork
-      git remote add apache https://git-wip-us.apache.org/repos/asf/$x
-      popd
-    done
+hub fork
+# the above triggers a fork at GitHub and then does a git remote rename of origin to upstream 
+# and a git add origin pointing to your fork
```
Comment (rdowner) [inline guide/dev/code/submodules.md]: This is not accurate. `hub fork`won't rename or remove any remotes, but it does add a new remote with your own username pointing to the new fork:

```bash
# git clone https://github.com/apache/brooklyn.git
[...]
# cd brooklyn
# git remote -v
origin  https://github.com/apache/brooklyn.git (fetch)
origin  https://github.com/apache/brooklyn.git (push)
# hub fork
[...]
# git remote -v
origin  https://github.com/apache/brooklyn.git (fetch)
origin  https://github.com/apache/brooklyn.git (push)
rdowner https://github.com/rdowner/brooklyn.git (fetch)
rdowner https://github.com/rdowner/brooklyn.git (push)
```

File: guide/dev/code/submodules.md [hunk 5]
```diff
@@ -5,31 +5,55 @@ layout: website-normal
...
+
+# configure upstream so pushes go to origin
+git remote set-url --push upstream $(git remote -v | grep origin | grep push | awk '{print $2}')
```
Comment (rdowner) [inline guide/dev/code/submodules.md]: Hacking the `upstream` push URL to actually push to a different repository? I've never seen that before, and I don't think I like it - it's quite deceptive :confused:

Is this article on recently-added Git support for triangular workflows helpful? https://github.com/blog/2042-git-2-5-including-multiple-worktrees-and-triangular-workflows

**Comment by rdowner** [comment]
After trying out the script a couple of times, IMO trying to script it is the wrong way to go about things. There are assumptions and edge cases which can trip it up. It also unnecessarily imposes a naming convention on our developers. Can we simply come up with a couple of one-liner commands which handles the basic case and let the power users figure out the rest themselves?

e.g. the basic clone-and-get-submodules in the "Get the code" page does a very effective job. Its only issue is that the remote is named `origin` when `upstream` would be more accurate. So maybe suggest that the user runs:
```bash
git remote rename origin upstream; git submodule foreach 'git remote rename origin upstream'
```

To fork everything, you just run this command:
```bash
hub fork; git submodule foreach 'hub fork'
```
This leaves you with a new remote that is your GH user ID, which is sensible. A power user who prefers a different name will know how to rename it, if that is what they want.

And then, default to pushing to your own repo:
```bash
git config remote.pushDefault mygithubid; git submodule foreach 'git config remote.pushDefault mygithubid'
```

I think this handful of one-liners will cover 75% of developers; it means I can do this:
```bash
# git branch my-new-feature-branch upstream/master
# git checkout my-new-feature-branch
# [make some commits]
# git push
To https://github.com/rdowner/brooklyn.git
 * [new branch]      my-new-feature-branch -> my-new-feature-branch
```
Note how my branch is tracking `upstream/master` for the purpose of `git pull`, but a `git push` goes to my own fork. (The remaining 25% of developers will be power users who know what they are doing!)

All the stuff about the GitHub PR branches and adding the Apache pure-git repo is unnecessary for almost everyone so can be omitted. Committers *do* need to know it, but that is what the [Committer Guide](https://brooklyn.apache.org/developers/committers/index.html) is for :smile:

**Comment by ahgittin** [comment]
@rdowner i strongly agree with almost everything.  i will:

* move github instructions to the unversioned "Developers" section 
* provide line-by-line instructions rather than a brittle magic script

the only thing i'm unsure of is PR's -- people other than committers may want to review PR's.  however i'll look for a good way to handle that.

**Comment by neykov** [comment]
@richardcloudsoft I am testing out your notes, one minor issue - `hub fork` requires the presence of an origin remote, so should be executed before the rename.

**Comment by ahgittin** [comment]
@neykov @richardcloudsoft see https://github.com/apache/brooklyn-docs/pull/11 -- i do the `hub fork` first

**Pull request merged**